component/kafka/group/component.go

```diff
@@ -0,0 +1,416 @@
+func consumerErrorsInc(name string) {
+	consumerErrors.WithLabelValues(name).Inc()
+}
+
+// topicPartitionOffsetDiffGaugeSet creates a new Gauge that measures partition offsets.
+func topicPartitionOffsetDiffGaugeSet(group, topic string, partition int32, high, offset int64) {
+	topicPartitionOffsetDiff.WithLabelValues(group, topic, strconv.FormatInt(int64(partition), 10)).Set(float64(high - offset))
+}
+
+// messageStatusCountInc increments the messageStatus counter for a certain status.
+func messageStatusCountInc(status, group, topic string) {
+	messageStatus.WithLabelValues(status, group, topic).Inc()
+}
+
+// New initializes a new  kafka consumer component with support for functional configuration.
+// The default failure strategy is the ExitStrategy.
+// The default batch size is 1 and the batch timeout is 100ms.
+// The default number of retries is 0 and the retry wait is 0.
+func New(name, group string, brokers, topics []string, proc kafka.BatchProcessorFunc, oo ...OptionFunc) (*Component, error) {
+	var errs []error
+	if name == "" {
+		errs = append(errs, errors.New("name is required"))
+	}
+
+	if group == "" {
+		errs = append(errs, errors.New("consumer group is required"))
+	}
+
+	if validation.IsStringSliceEmpty(brokers) {
+		errs = append(errs, errors.New("brokers are empty or have an empty value"))
+	}
+
+	if validation.IsStringSliceEmpty(topics) {
+		errs = append(errs, errors.New("topics are empty or have an empty value"))
+	}
+
+	if proc == nil {
+		errs = append(errs, errors.New("work processor is required"))
+	}
+
+	if len(errs) > 0 {
+		return nil, patronErrors.Aggregate(errs...)
+	}
+
+	defaultSaramaCfg, err := kafka.DefaultSaramaConfig(name)
+	if err != nil {
+		return nil, err
+	}
+
+	cmp := &Component{
+		name:         name,
+		group:        group,
+		brokers:      brokers,
+		topics:       topics,
+		proc:         proc,
+		retries:      defaultRetries,
+		retryWait:    defaultRetryWait,
+		batchSize:    defaultBatchSize,
+		batchTimeout: defaultBatchTimeout,
+		failStrategy: defaultFailureStrategy,
+		saramaConfig: defaultSaramaCfg,
+	}
+
+	for _, optionFunc := range oo {
+		err = optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Component is a kafka consumer implementation that processes messages in batch
+type Component struct {
+	name         string
+	group        string
+	topics       []string
+	brokers      []string
+	saramaConfig *sarama.Config
+	proc         kafka.BatchProcessorFunc
+	failStrategy kafka.FailStrategy
+	batchSize    uint
+	batchTimeout time.Duration
+	retries      uint
+	retryWait    time.Duration
+	commitSync   bool
+}
+
+// Run starts the consumer processing loop to process messages from Kafka.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	return c.processing(ctx)
+}
+
+func (c *Component) processing(ctx context.Context) error {
+	var componentError error
+
+	retries := int(c.retries)
```

Comment:
Yea I thought about it too. Let me think about a solution. Perhaps the handler can feedback whether it had any successful processing...

Reply:
Done 👍 .. if a message is processed successfully, the retry counter is set again to zero.

Reply:
Just one point here - as suggested in [this](https://github.com/beatlabs/patron/issues/266) issue there could be other retry strategies worth considering.

One option would be to use https://github.com/cenkalti/backoff allowing it to be overridden via configuration.
Ideally with separate settings for processorFunc and connection errors, if possible.

This can be done later but it might be good to think about how the current design can accommodate something like this in the future.

Reply:
> Only consider sequential errors (any successfully processed message resets the retry counter)
Is actually handled here so this part of the issue is resolved in this component.

Regarding the rest, very good points and we actually are able here to differentiate between processor/handler errors and connection errors so I think the component can definitely be further improved in that direction. 

That said, I think this PR has already seen a lot of back and forth so I would like to handle the issue you raised in a separate PR after this is merged if you don't mind.

Reply:
I agree. Let's create a new ticket to introduce exponential back off to all new consumer we build (kafka, AMQP, SQS).